### PR TITLE
[NP-11036] UserLifecycleTicket

### PR DESCRIPTION
**Pull request by jlhughes**
### System to delete a user and all associated data.
LifecycleAware data is marked `DELETED`
EnabledAware data is marked `false`
otherwise a regular DAO.remove

UserLifecycleTicket will mark the user `DELETED`, but also all relationship data - such as accounts, contacts, credentials, ... 
1. To ensure the system is consistent - if the user is deleted, then it's associations are of no use to the system. (Remembering that things are just marked deleted, and always available if we really need them). 
2. Ensure associated referralCodes, UCJs, accounts, for example, can't be used independent of the user. 
3. Support GDPR with regard to data deletion (being normally inaccessible). 

### Other
- On failure of `DELETED`, user will be `DISABLED`. 
- The ticket stores a report of what was updated, with the intent that it can be used to un-disable or un-delete. 
- User can be `DISABLED` or `DELETED` then later `ACTIVE` again using the previous `update` list.

### CODE REVIEW

File: src/foam/nanos/auth/User.js [hunk 1]
```diff
@@ -1024,17 +1044,91 @@ foam.CLASS({
...
+    },
+    {
+      name: 'activateUser',
+      label: 'Activate',
+      toolTip: 'Open a ticket to activate a new user or previously disabled user',
+      availablePermissions: ['user.action.activate'],
+      isAvailable: async function(id, spid, lifecycleState) {
+        // NOTE: testing spid as hack so action only available from detail view
+        return id && spid && ( lifecycleState == this.LifecycleState.DISABLED ||
+                               lifecycleState == this.LifecycleState.PENDING );
```
Comment (lchanmann) [inline src/foam/nanos/auth/User.js]: should it also be shown for deleted lifecycle state user?

File: src/foam/nanos/auth/ruler/UserLifecycleTicketRuleAction.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      name: 'applyAction',
+      javaCode: `
+        agency.submit(x, new ContextAgent() {
+          @Override
+          public void execute(X x) {
+            X rulerX = ruler.getX();
+            User admin = new User();
+            admin.setId(1);
+            admin.setGroup("admin");
+            X systemX = foam.util.Auth.sudo(ruler.getX(), admin);
```
Comment (lchanmann) [inline src/foam/nanos/auth/ruler/UserLifecycleTicketRuleAction.js]: rulerX should already be the system context, do we still need to sudo as admin?

File: src/foam/nanos/auth/ruler/UserLifecycleTicketRuleAction.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                user.setLifecycleState(nu);
+                ((Logger) x.get("logger")).info("UserLifecycleTicket", user.getId(), old, nu);
+                ((DAO) ruler.getX().get("localUserDAO")).put(user);
+              }
+              ticket.setMessage(null);
+            } catch (Throwable t) {
+              ticket.setMessage(t.getMessage());
+              if ( nu == LifecycleState.DISABLED ||
+                   nu == LifecycleState.DELETED ) {
+                user.setLifecycleState(LifecycleState.DISABLED);
```
Comment (lchanmann) [inline src/foam/nanos/auth/ruler/UserLifecycleTicketRuleAction.js]: should the user lifecycle state be set according to the requested cycle state of the ticket?

File: src/foam/nanos/auth/ruler/UserLifecycleTicketSink.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  ],
+
+  methods: [
+    {
+      name: 'put',
+      javaCode: `
+      X x = getX();
+      Logger logger = (Logger) x.get("logger");
+      PropertyInfo prop = (PropertyInfo) ((FObject)obj).getClassInfo().getAxiomByName("id");
+      Object id = prop != null ? prop.get(obj) : null;
```
Comment (lchanmann) [inline src/foam/nanos/auth/ruler/UserLifecycleTicketSink.js]: FObjects that has "id" property are implementing foam.core.Identifiable interface. We may get the primary from it directly.
```suggestion
      Object id = obj instanceof Identifiable ? ((Identifiable) obj).getPrimaryKey() : null;
```

File: src/foam/nanos/auth/User.js (hunk 1 above)
Comment (jlhughes) [reply src/foam/nanos/auth/User.js]: Not yet supported.  un-disable isn't really either, but works enough now to be useful.

File: src/foam/nanos/auth/ruler/UserLifecycleTicketRuleAction.js (hunk 3 above)
Comment (jlhughes) [reply src/foam/nanos/auth/ruler/UserLifecycleTicketRuleAction.js]: No, this is fallback when the updateUserAssociations throws an exception when the user can't or shouldn't be changed - for example, in nanopay, if there are active transactions, we can't delete, only disable.  So regardless of what was requested, we at least disable the user so they can't log in.

File: src/foam/nanos/auth/ruler/UserLifecycleTicketRuleAction.js (hunk 2 above)
Comment (jlhughes) [reply src/foam/nanos/auth/ruler/UserLifecycleTicketRuleAction.js]: removed.

**Review by lchanmann — APPROVED** [review]

File: src/foam/nanos/auth/ruler/UserLifecycleTicket.js [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  ],
+
+  properties: [
+    {
+      name: 'status',
+      order: 5,
+      createVisibility: 'HIDDEN'
+    },
+    {
+      name: 'statusChoices',
```
Comment (adamfox-nanopay) [inline src/foam/nanos/auth/ruler/UserLifecycleTicket.js:41]: isnt this inherited from Ticket?

**Pull request merged**